### PR TITLE
Issue 15561 adding contentlets to json response

### PR DESCRIPTION
**Pull request by wezell**
This pull request adds the contentlet list to each container in our PageResource `/render/` endpoint.  It also adds the content list to the `/json/` endpoint.  It also adds the container and content list to non-layout driven pages.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 1]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
 
+        final Table<String, String, Set<String>> pageContents = APILocator.getMultiTreeAPI().getPageMultiTrees(htmlPage, live);
+        final List<ContainerRaw> raws = Lists.newArrayList();
         if (!pageContents.isEmpty()) {
 
             for (final String containerId : pageContents.rowKeySet()) {
-                for (final String uniqueId : pageContents.row(containerId)
-                    .keySet()) {
+                final Container container =
+                        (live && APILocator.getVersionableAPI().findLiveVersion(containerId, APILocator.systemUser(), false) != null)
```
[the remark this answers is not in the document]
Comment (wezell) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: I think we need to change the versionable to either not accept containers or to do the work if the identifier passed in is a "container".  I don't know what that looks like.

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 2]
```diff
@@ -25,51 +45,49 @@
...
-    final Date timeMachine;
-    List<Tag> pageFoundTags;
+    final List<Tag> pageFoundTags;
     final StringBuilder widgetPreExecute;
     final static String WIDGET_PRE_EXECUTE = "WIDGET_PRE_EXECUTE";
+    final List<ContainerRaw> containersRaw;
+    final long languageId;
 
-    public PageContextBuilder(IHTMLPage htmlPage, User user, PageMode mode, Date timeMachine)
+    public PageContextBuilder(IHTMLPage htmlPage, User user, PageMode mode, long languageId)
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 3]
```diff
@@ -25,51 +45,49 @@
...
-    }
-
+        this.ctxMap = populateContext();
+        this.containersRaw = populateContainers();
 
+    }
 
-    public PageContextBuilder(IHTMLPage htmlPage, User user, PageMode mode) throws DotSecurityException, DotDataException {
-        this(htmlPage, user, mode, null);
+    public PageContextBuilder(IHTMLPage htmlPage, User user, PageMode mode)
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: I can do that, throws an exception? InvalidArgument or so?

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 4]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
+                                ? (Container) APILocator.getVersionableAPI().findLiveVersion(containerId, APILocator.systemUser(), false)
+                                : (Container) APILocator.getVersionableAPI().findWorkingVersion(containerId, APILocator.systemUser(),
+                                        false);
+
+
+                if (container == null) {
+                    continue;
+                }
+                List<ContainerStructure> containerStructures = APILocator.getContainerAPI().getContainerStructures(container);
+                Map<String, List<Contentlet>> contentMaps = Maps.newLinkedHashMap();
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: Done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 5]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
-                    final User systemUser = APILocator.getUserAPI().getSystemUser();
-                    final Container container = (live && APILocator.getContainerAPI().getLiveContainerById(containerId, systemUser, false) !=null)
-                            ? APILocator.getContainerAPI().getLiveContainerById   (containerId, systemUser, false)
-                            : APILocator.getContainerAPI().getWorkingContainerById(containerId, systemUser, false);
 
-                    if(container==null)continue;
-                    boolean hasWritePermissionOnContainer =
-                            permissionAPI.doesUserHavePermission(container, PERMISSION_WRITE, user, false) && APILocator.getPortletAPI()
-                                .hasContainerManagerRights(user);
+                    boolean hasWritePermissionOnContainer = permissionAPI.doesUserHavePermission(container, PERMISSION_WRITE, user, false)
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: Done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 6]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
+                    }
                     // to check user has permission to write this container
                     boolean hasWritePermOverTheStructure = false;
 
-                    for (ContainerStructure cs : APILocator.getContainerAPI()
-                        .getContainerStructures(container)) {
-                        Structure st = CacheLocator.getContentTypeCache()
-                            .getStructureByInode(cs.getStructureId());
-
+                    for (ContainerStructure cs : containerStructures) {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: Done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 7]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
+                        Structure st = CacheLocator.getContentTypeCache().getStructureByInode(cs.getStructureId());
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: done, renamed to structure

[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: Done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 8]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
-                                    catch(Exception e) {
-                                        throw new DotStateException(e);
-                                    }
-                                }
-                            )
-                            .filter(Objects::nonNull)
-                            .collect(Collectors.toList());
-                    
-                    // get contentlets only for main frame
+                    List<Contentlet> contentlets = cons.stream().map(id -> {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: Done

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 9]
```diff
@@ -141,79 +155,78 @@ private void populateContext() throws DotDataException, DotSecurityException {
...
+                        try {
+                            Contentlet con = APILocator.getContentletAPI().findContentletForLanguage(this.languageId, APILocator.getIdentifierAPI().find(id));
+                            return (con!=null) ? con : APILocator.getContentletAPI().findContentletByIdentifierAnyLanguage(id);
+                        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: not so much to do

File: dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java [hunk 10]
```diff
@@ -234,68 +246,61 @@ private void populateContainers() throws DotDataException, DotSecurityException
                         }
                     }
 
-                    
-                    
-                    String[] contentlist = contentlets.stream()
-                            .map(con -> con.getIdentifier())
-                            .toArray(size -> new String[size]);
+
+                    String[] contentlist = contentlets.stream().map(con -> con.getIdentifier()).toArray(size -> new String[size]);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java [hunk 11]
```diff
@@ -148,6 +149,8 @@ private void addEmptyContainers(final IHTMLPage page, Table<String, String, Set<
                                     final boolean liveMode) throws DotDataException, DotSecurityException {
 
         try {
+            Template template = APILocator.getTemplateAPI().findWorkingTemplate(page.getTemplateId(), APILocator.getUserAPI().getSystemUser(), false);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java [hunk 12]
```diff
@@ -148,6 +149,8 @@ private void addEmptyContainers(final IHTMLPage page, Table<String, String, Set<
...
+            if(!template.isDrawed()) return;
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/factories/MultiTreeAPIImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java [hunk 13]
```diff
@@ -27,6 +27,7 @@
 import com.dotmarketing.util.Constants;
 import com.dotmarketing.util.UtilMethods;
 import com.dotmarketing.util.WebKeys;
+import com.google.common.collect.Lists;
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/containers/business/ContainerAPIImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java [hunk 14]
```diff
@@ -1,108 +1,74 @@
...
+    public Collection<? extends ContainerRaw> getContainers(final HTMLPageAsset page, final PageMode mode)
             throws DotSecurityException, DotDataException {
-        return getContainersRendered(template, null, mode);
+        return getContainersRendered(page, null, mode);
     }
 
-    public Collection<ContainerRendered> getContainersRendered(final Template template, final Context velocityContext,
-                                                         final PageMode mode)
 
+    public Collection<? extends ContainerRaw> getContainersRendered(HTMLPageAsset page, final Context velocityContext, PageMode mode)
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java [hunk 15]
```diff
@@ -1,108 +1,74 @@
...
             throws DotSecurityException, DotDataException {
 
-        if (template != null && !template.isDrawed()) {
-            return Collections.EMPTY_LIST;
-        }
+   
+        PageContextBuilder pcb = new PageContextBuilder(page, APILocator.systemUser(), mode);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java [hunk 16]
```diff
@@ -1,108 +1,74 @@
...
-                final List<ContainerStructure> containerStructures = this.containerAPI.getContainerStructures(container);
 
-                Map<String, String> containersRendered = velocityContext != null ?
-                        render(velocityContext, mode, entry.getValue(), container): null;
 
-                return new ContainerRendered(container, containerStructures, containersRendered);
-            } catch (NotFoundInDbException e) {
+        return pcb.getContainersRaw().stream().map(cRaw -> {
+            try {
+                Map<String, String> uuidsRendered = render(velocityContext, mode, cRaw);
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java [hunk 17]
```diff
@@ -1,108 +1,74 @@
...
-
-                    try {
-                        return VelocityUtil.getInstance().mergeTemplate(key.path, velocityContext);
-                    } catch (Exception e) {
-                        Logger.warn(this,String.format("Container '%s' could not be " +
-                                "rendered via " + "Velocity.", container.getIdentifier()), e);
-                        return StringPool.BLANK;
-                    }
-                }));
+    private Map<String, String> render(Context velocityContext, PageMode mode, ContainerRaw cRaw) {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java [hunk 18]
```diff
@@ -1,108 +1,74 @@
...
+
+        Map<String, String> rendered = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java [hunk 19]
```diff
@@ -1,108 +1,74 @@
...
+        for (String uuid : cRaw.getContentlets().keySet()) {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/ContainerRenderedBuilder.java]: Done

**Review by jgambarios — APPROVED** [review]

**Pull request merged**